### ISSUE TITLE
Support & maintain a cloudformation based setup for Kubernetes on AWS

### ISSUE THREAD

**Issue by chrisleck** [kubernetes/kubernetes#21275]
I have a fully declarative setup for our Kubernetes clusters using cloudformation, written in jsonnet, which I've discussed with a few contributors here.  I'd love to contribute this and hook it into ongoing updates to the cluster bootstrap process.

Two questions:
1. Is this something that would be welcome?
2. Where would it live?

**Comment by kevinjcash** [comment]
I would like this very much, and I can contribute if you've got something started. I don't see why it couldn't live in the example directory at the very least, if not cluster/aws.

**Comment by a-robinson** [comment]
cc @brendandburns

**Comment by aronchick** [comment]
HUGELY interesting - please let us know how to get it in!

**Comment by chrisleck** [comment]
Just let me know where this should go in the source tree, and I'll sanitize my company-specific pieces out and submit a PR.

**Comment by aronchick** [comment]
cc @justinsb Can you make a suggestion where this should live?

**Comment by justinsb** [comment]
TLDR: `cluster/aws-cloudformation`

TL:

I think this is wonderful, and part of a bigger conversation about how we plan on moving to a post-bash/Salt world.

It would naturally be welcome in https://github.com/kubernetes/contrib .  The question in my mind is whether we can do better.

I'd almost say @aronchick that this might be a question for you - it's not really a technical question.  Here's how I see it:

- Everyone dislikes the kube-up/salt stuff we have now, but it produces a working cluster for the common case
- Everyone would like an alternative, but there's no clear preference.  CloudFormation doesn't work on anything but AWS.  Lots of AWS people hate CloudFormation.  Some people will want to use Terraform.  Others Chef, Puppet, Ansible etc.  Do we accept all of these into core?  Do we remove kube-up from core?
- All the alternatives so far (and I'm not talking about this proposal) have pretty major shortcomings (e.g. TLS certificates), and cause us a lot of support problems and give a generally poor user experience.
- Does accepting something into core mean that certain standards are required of it?  For example, should it pass e2e testing?


I would love to have something in core that people actually want to use.  I want to tell people "not building from kube-up is like building Linux from scratch - you have to know what you're doing and can't expect a lot of help", but that's just not realistic when there are so many valid reasons not to use kube-up.  I personally think we should split kubernetes-users into people that used kube-up or another approved installation method, and people that did not or are trying to self-install.  But to do that we have to have an installation method we're proud of, and I can't in good faith say that is kube-up (even as someone that spends a lot of time tending AWS kube-up).

I'd love to see a PR for this - having a PR forces us to discuss it.  I don't know if there is a SIG that is discussing this, but if so we should involve them.

If it doesn't actually share anything with the current `cluster/aws` tree, we could put it under `cluster/aws-cloudformation` or something like that.  So it would work with `export KUBERNETES_PROVIDER=aws-cloudformation`  That would be entirely opt-in, so I'd have no problems with that.  We have other providers in there that are pretty broken, so I don't see the harm in incubating this there.  It also means we don't have to worry about breaking anything.

I think generally by making k8s more self managing (though k8s on k8s, NodeSets etc) we can reduce the amount of work that needs to be done by the installer script, to the point where it becomes trivial to bootstrap.  At that point I hope we can have the "big tent" approach where we do have cloudformation, terraform etc templates in tree, but at the same time they would become trivial.  They're not there yet, but maybe it's time to bring in a CF template, so we can start to see what needs to be done and how we can start to trim it down.

**Comment by aronchick** [comment]
IMHO, I believe a lot of this comes down to breaking out cloud provider to make it easier, and breaking cloud providers out of core. @brendandburns has some thoughts on this.

But, yes, we need to have solutions for the largest clouds today. In order of priority, i'd say:

1. Terraform
2. Cloud Formation
3. Deployment Manager
4. Chef/Puppet
5. Everything else

**Comment by chrisleck** [comment]
I've started a port of what we use internally.  It currently does the following, which seems like a pretty bare bones but production-ready cluster, though I'm sure we'll have a fun discussion once I submit the PR for review :)

* Sets up IAM roles for certificate administration, etcd, kube-master, and kube-node instances
* Sets up a dedicated and locked-down S3 bucket for hosting certificates to be distributed to instances - still wish there was a better way :)
* Creates a regional VPC, with public and private subnets & routing tables, internet gateway, NAT instances, public and private DNS zones, and security groups for the NAT instances and a break-glass group for direct SSH access
* Sets up SNS for firing alerts
* Creates regional infrastructure, including CloudTrail for tracking cluster changes, and setting up re-usable security groups for KubeMaster & KubeNode
* Creates a multi-AZ etcd cluster on CoreOS using internal DNS service records for bootstrapping.
* Per active AZ, sets up a CoreOS/flanneld-based Kubernetes cluster with a single master instance and N auto-scaler driven nodes; exposing the master to defined corp IP blocks via a public DNS address. Uses server & client certificates for authn, driven from something like TinyCert.  Master and nodes self-bootstrap via cloud-config.
* Sets up addons per cluster (dns, kube-dash, kube-ui, kibana, influxdb, grafana) and exposes them via public DNS to defined corp IP blocks.

**Comment by kevinjcash** [comment]
Thanks @chrisleck, looking forward to testing this.

**Comment by chrisleck** [comment]
https://github.com/kubernetes/kubernetes/pull/22227
